src/nnue/nnue_feature_transformer.h

```diff
@@ -651,8 +665,157 @@ class FeatureTransformer {
+                auto columnPsqt        = reinterpret_cast<const psqt_vec_t*>(&psqtWeights[offset]);
+
+                for (std::size_t k = 0; k < NumPsqtRegs; ++k)
+                    psqt[k] = vec_add_psqt_32(psqt[k], columnPsqt[k]);
+            }
+            for (int i = 0; i < int(removed.size()); ++i)
+            {
+                IndexType       index  = removed[i];
+                const IndexType offset = PSQTBuckets * index + j * PsqtTileHeight;
+                auto columnPsqt        = reinterpret_cast<const psqt_vec_t*>(&psqtWeights[offset]);
+
+                for (std::size_t k = 0; k < NumPsqtRegs; ++k)
+                    psqt[k] = vec_sub_psqt_32(psqt[k], columnPsqt[k]);
+            }
+
+            for (std::size_t k = 0; k < NumPsqtRegs; ++k)
+                vec_store_psqt(&entryTilePsqt[k], psqt[k]);
+        }
+
+#else
+
+        for (const auto index : added)
+        {
+            const IndexType offset = HalfDimensions * index;
+            for (IndexType j = 0; j < HalfDimensions; ++j)
+                entry.acc.accumulation[Perspective][j] += weights[offset + j];
+
+            for (std::size_t k = 0; k < PSQTBuckets; ++k)
+                entry.acc.psqtAccumulation[Perspective][k] +=
+                psqtWeights[index * PSQTBuckets + k];
+        }
+        for (const auto index : removed)
+        {
+            const IndexType offset = HalfDimensions * index;
+            for (IndexType j = 0; j < HalfDimensions; ++j)
+                entry.acc.accumulation[Perspective][j] -= weights[offset + j];
+
+            for (std::size_t k = 0; k < PSQTBuckets; ++k)
+                entry.acc.psqtAccumulation[Perspective][k] -=
+                psqtWeights[index * PSQTBuckets + k];
+        }
+
+#endif
+
+        // The accumulator of the refresh entry has been updated.
+        // Now copy its content to the actual accumulator we were refreshing
+
+        std::memcpy(accumulator.psqtAccumulation[Perspective],
+                    entry.acc.psqtAccumulation[Perspective],
+                    sizeof(int32_t) * PSQTBuckets);
+
+        std::memcpy(accumulator.accumulation[Perspective],
+                    entry.acc.accumulation[Perspective],
+                    sizeof(int16_t) * HalfDimensions);
+
+        for (int i = WHITE; i <= BLACK; i++)
+            entry.byColorBB[Perspective][i] = pos.pieces(Color(i));
+
+        for (int i = PAWN; i <= KING; i++)
+            entry.byTypeBB[Perspective][i] = pos.pieces(PieceType(i));
+    }
+
     template<Color Perspective>
     void update_accumulator_refresh(const Position& pos, bool psqtOnly) const {
+
+        // When we are refreshing the accumulator of the big net, 
+        // redirect to the version of refresh that uses the refresh table
+        if (HalfDimensions == Eval::NNUE::TransformedFeatureDimensionsBig) {
+            // TODO: find a better solution than const_casting the position
```

Comment:
Could we perhaps implement the same for the small network and get rid of the old update code completely?

I assume the const_cast is needed due to modifications of `refresh_table`. In such cases it's normal to make the cache fields `mutable`. This ofc makes the class not thread-safe (not less so than const_cast though), but I don't think we use this in any concurrent way anyway.

Reply:
After this gets committed I'm happy to help make it work for the small network in a follow up PR.

Reply:
already at it https://tests.stockfishchess.org/tests/view/662557c33fe04ce4cefc3906

Reply:
see also https://tests.stockfishchess.org/tests/view/662557c33fe04ce4cefc3906